sympy/solvers/solvers.py

```diff
@@ -1956,12 +1957,17 @@ def solve_linear(lhs, rhs=0, symbols=[], exclude=[]):
                                 i.function.is_number]
                         # do a slight bit of simplification
                         vi = expand_mul(vi.subs(irep))
-                        if not d.has(xi) or not (d/xi).has(xi):
+                        if xi not in dfree:
+                            return xi, vi
+                        if xi not in (d/xi).free_symbols:
```

Comment:
think more about this...seems like this is equivalent to `xi not in dn.free_symbols` which is already the condition of this `if-block`. Perhaps both of these tests are not necessary and `return xi, vi` is all that is needed. Then factor_terms would not be needed.

[ ] check to see what (if anything) gets by these two tests.